scala_proto/private/scalapb_aspect.bzl

```diff
@@ -0,0 +1,230 @@
+load(
+    "//scala/private:common.bzl",
+    "write_manifest_file",
+)
+load("//scala/private:rule_impls.bzl", "compile_scala")
+
+load("//scala_proto/private:proto_to_scala_src.bzl", "proto_to_scala_src")
+
+
+ScalaPBAspectInfo = provider(fields = [
+    "proto_info",
+    "src_jars",
+    "output_files",
+    "java_info",
+])
+
+
+ScalaPBImport = provider(fields = [
+    "java_info",
+    "proto_info",
+])
+
+
+ScalaPBInfo = provider(fields = [
+    "aspect_info",
+])
+
+
+def merge_proto_infos(tis):
+    return struct(
+        transitive_sources = [t.transitive_sources for t in tis],
+    )
+
+def merge_scalapb_aspect_info(scalapbs):
+    return ScalaPBAspectInfo(
+        src_jars = depset(transitive = [s.src_jars for s in scalapbs]),
+        output_files = depset(transitive = [s.output_files for s in scalapbs]),
+        proto_info = merge_proto_infos([s.proto_info for s in scalapbs]),
+        java_info = java_common.merge([s.java_info for s in scalapbs]),
+    )
+
+
+def _compiled_jar_file(actions, scalapb_jar):
+    scalapb_jar_name = scalapb_jar.basename
+
+    # ends with .srcjar, so remove last 6 characters
+    without_suffix = scalapb_jar_name[0:len(scalapb_jar_name) - 6]
+
+    # this already ends with _scalapb because that is how scalapb_jar is named
+    compiled_jar = without_suffix + "jar"
+    return actions.declare_file(compiled_jar, sibling = scalapb_jar)
+
+def _compile_scala(
+        ctx,
+        scalac,
+        label,
+        output,
+        scalapb_jar,
+        deps_java_info,
+        implicit_deps):
+    manifest = ctx.actions.declare_file(
+        label.name + "_MANIFEST.MF",
+        sibling = scalapb_jar,
+    )
+    write_manifest_file(ctx.actions, manifest, None)
+    statsfile = ctx.actions.declare_file(
+        label.name + "_scalac.statsfile",
+        sibling = scalapb_jar,
+    )
+    merged_deps = java_common.merge(deps_java_info + implicit_deps)
+
+    # this only compiles scala, not the ijar, but we don't
+    # want the ijar for generated code anyway: any change
+    # in the proto generally will change the interface and
+    # method bodies
+    compile_scala(
+        ctx,
+        Label("%s-fast" % (label)),
```

Comment:
starlark really needs some kind of gensym thing that makes sure generated symbols are unique.